# CELLEX

In [1]:
import pandas as pd
import cellex

In [2]:
data_dir = 'data/CELLEX'
file_s = data_dir+'/Allen_Human_Middle_temporal_gyrus/human_MTG_2018-06-14_samples-columns.csv'
file_g = data_dir+'/Allen_Human_Middle_temporal_gyrus/human_MTG_2018-06-14_genes-rows.csv'
file_e = data_dir+'/Allen_Human_Middle_temporal_gyrus/human_MTG_2018-06-14_exon-matrix.csv'
file_i = data_dir+'/Allen_Human_Middle_temporal_gyrus/human_MTG_2018-06-14_intron-matrix.csv'

In [3]:
%%time
samples = pd.read_csv(file_s)
genes = pd.read_csv(file_g)
exon = pd.read_csv(file_e)
intron = pd.read_csv(file_i)

exon.drop('Unnamed: 0',axis=1, inplace=True)
intron.drop('Unnamed: 0',axis=1, inplace=True)

CPU times: user 6min 16s, sys: 3min 35s, total: 9min 51s
Wall time: 13min 50s


In [4]:
display(genes)

gene chromosome  entrez_id  \
0          3.8-1.2          6     353007   
1          3.8-1.3          6     353008   
2          3.8-1.4          6     353009   
3          3.8-1.5          6     353010   
4          5-HT3C2          3     389180   
...            ...        ...        ...   
50276          ZYX          7       7791   
50277        ZZEF1         17      23140   
50278         ZZZ3          1      26009   
50279   bA255A11.4          9     259265   
50280  bA395L14.12          2     150478   

                                               gene_name mouse_homologenes  
0      HLA complex group 26 (non-protein coding) pseu...               NaN  
1      HLA complex group 26 (non-protein coding) pseu...               NaN  
2      HLA complex group 26 (non-protein coding) pseu...               NaN  
3      HLA complex group 26 (non-protein coding) pseu...               NaN  
4      5-hydroxytryptamine (serotonin) receptor 3, fa...               NaN  
...                                                  ...               ...  
50276                                              zyxin               Zyx  
50277         zinc finger, ZZ-type with EF-hand domain 1             Zzef1  
50278                  zinc finger, ZZ-type containing 3              Zzz3  
50279                        melanoma antigen pseudogene               NaN  
50280  small nuclear ribonucleoprotein polypeptide A'...               NaN  

[50281 rows x 5 columns]

In [5]:
data = exon+intron
data.index = genes.gene
display(data)

F1S4_160106_001_B01  F1S4_160106_001_C01  F1S4_160106_001_E01  \
gene                                                                         
3.8-1.2                        0                    0                    0   
3.8-1.3                        0                    0                    0   
3.8-1.4                        0                    0                    0   
3.8-1.5                        0                    0                    0   
5-HT3C2                        0                    0                    0   
...                          ...                  ...                  ...   
ZYX                            0                    0                   72   
ZZEF1                          1                    8                  197   
ZZZ3                         306                  341                  296   
bA255A11.4                     0                    0                    0   
bA395L14.12                    0                    1                    1   

             F1S4_160106_001_G01  F1S4_160106_001_H01  F1S4_160106_002_C01  \
gene                                                                         
3.8-1.2                        0                    0                    0   
3.8-1.3                        0                    0                    0   
3.8-1.4                        0                    0                    0   
3.8-1.5                        0                    0                    0   
5-HT3C2                        0                    0                    0   
...                          ...                  ...                  ...   
ZYX                            0                   10                    0   
ZZEF1                         58                  169                   31   
ZZZ3                         206                    0                  202   
bA255A11.4                     0                    0                    0   
bA395L14.12                    1                    1                    0   

             F1S4_160106_002_D01  F1S4_160106_002_F01  F1S4_160106_002_G01  \
gene                                                                         
3.8-1.2                        0                    0                    0   
3.8-1.3                        0                    0                    0   
3.8-1.4                        0                    0                    0   
3.8-1.5                        0                    0                    0   
5-HT3C2                        0                    0                    0   
...                          ...                  ...                  ...   
ZYX                            0                   19                    0   
ZZEF1                        200                    0                    0   
ZZZ3                         378                  110                    5   
bA255A11.4                     0                    0                    0   
bA395L14.12                    0                    0                    0   

             F1S4_160106_002_H01  ...  F2S4_170405_059_E01  \
gene                              ...                        
3.8-1.2                        0  ...                    0   
3.8-1.3                        0  ...                    0   
3.8-1.4                        0  ...                    0   
3.8-1.5                        0  ...                    0   
5-HT3C2                        0  ...                    0   
...                          ...  ...                  ...   
ZYX                            0  ...                    3   
ZZEF1                          1  ...                  130   
ZZZ3                         297  ...                   50   
bA255A11.4                     0  ...                    0   
bA395L14.12                    0  ...                    0   

             F2S4_170405_059_F01  F2S4_170405_059_G01  F2S4_170405_059_H01  \
gene                                                                         
3.8-1

In [6]:
metadata = samples[['sample_name','cluster']]
display(metadata)# cluster name cannot contain spaces, other CELLECT will return an error

sample_name                  cluster
0      F1S4_160106_001_B01      Inh L4-6 SST B3GAT2
1      F1S4_160106_001_C01      Exc L5-6 RORB TTC12
2      F1S4_160106_001_E01       Exc L5-6 FEZF2 ABO
3      F1S4_160106_001_G01  Exc L5-6 FEZF2 EFTUD1P1
4      F1S4_160106_001_H01       Exc L3-5 RORB ESR1
...                    ...                      ...
15923  F2S4_170405_060_B01       Exc L5-6 FEZF2 ABO
15924  F2S4_170405_060_C01    Exc L5-6 THEMIS FGF10
15925  F2S4_170405_060_E01      Exc L4-6 FEZF2 IL26
15926  F2S4_170405_060_F01      Inh L4-6 SST B3GAT2
15927  F2S4_170405_060_H01    Inh L5-6 SST MIR548F2

[15928 rows x 2 columns]

In [7]:
metadata['cluster'].replace(' ', '_', regex=True, inplace=True)
metadata.set_index('sample_name', inplace=True)
display(metadata)

cluster
sample_name                                 
F1S4_160106_001_B01      Inh_L4-6_SST_B3GAT2
F1S4_160106_001_C01      Exc_L5-6_RORB_TTC12
F1S4_160106_001_E01       Exc_L5-6_FEZF2_ABO
F1S4_160106_001_G01  Exc_L5-6_FEZF2_EFTUD1P1
F1S4_160106_001_H01       Exc_L3-5_RORB_ESR1
...                                      ...
F2S4_170405_060_B01       Exc_L5-6_FEZF2_ABO
F2S4_170405_060_C01    Exc_L5-6_THEMIS_FGF10
F2S4_170405_060_E01      Exc_L4-6_FEZF2_IL26
F2S4_170405_060_F01      Inh_L4-6_SST_B3GAT2
F2S4_170405_060_H01    Inh_L5-6_SST_MIR548F2

[15928 rows x 1 columns]

In [8]:
metadata.cluster.unique() # contains the cluster no_class

array(['Inh_L4-6_SST_B3GAT2', 'Exc_L5-6_RORB_TTC12', 'Exc_L5-6_FEZF2_ABO',
       'Exc_L5-6_FEZF2_EFTUD1P1', 'Exc_L3-5_RORB_ESR1',
       'Exc_L4-5_RORB_FOLH1B', 'Inh_L1-4_LAMP5_LCP2', 'Oligo_L1-6_OPALIN',
       'Exc_L5-6_THEMIS_C1QL3', 'Inh_L2-3_VIP_CASC6',
       'Exc_L4-6_RORB_SEMA3E', 'Inh_L4-6_PVALB_SULF1',
       'Exc_L5-6_THEMIS_CRABP1', 'Exc_L4-6_FEZF2_IL26',
       'Exc_L4-5_RORB_DAPK2', 'Exc_L3-5_RORB_TWIST2', 'Inh_L3-6_SST_HPGD',
       'Inh_L5-6_SST_NPM1P10', 'Inh_L5-6_SST_MIR548F2',
       'Inh_L4-5_PVALB_MEPE', 'Inh_L5-6_SST_TH', 'OPC_L1-6_PDGFRA',
       'Inh_L1-2_VIP_PCDH20', 'Inh_L1-2_PAX6_CDH12', 'Inh_L1_SST_NMBR',
       'Astro_L1-2_FGFR3_GFAP', 'Exc_L2-3_LINC00507_FREM3',
       'Exc_L2_LAMP5_LTK', 'Inh_L1-3_PAX6_SYT6', 'Inh_L1-2_SST_BAGE2',
       'Astro_L1-6_FGFR3_SLC14A1', 'Inh_L1-2_VIP_TSPAN12',
       'Inh_L1-3_VIP_CHRM2', 'Inh_L1-2_GAD1_MC4R', 'Inh_L1_SST_CHRNA4',
       'Micro_L1-3_TYROBP', 'Inh_L2-5_VIP_TYR', 'Inh_L1-3_VIP_GGH',
       'Inh_L2-6_LAMP5_CA1',

In [9]:
nan_cells = metadata[metadata['cluster']=='no_class'].index
len(nan_cells)

325

In [10]:
data_dropped = data.drop(nan_cells, axis=1)
print(data.shape, data_dropped.shape)

(50281, 15928) (50281, 15603)


In [11]:
metadata_dropped = metadata[metadata['cluster']!='no_class']
print(metadata.shape, metadata_dropped.shape)

(15928, 1) (15603, 1)


In [12]:
%%time
eso = cellex.ESObject(data=data_dropped,
                      annotation=metadata_dropped,
                      verbose=True)

Preprocessing - checking input ... input parsed in 0 min 0 sec
Preprocessing - running remove_non_expressed ... excluded 1914 / 50281 genes in 4 min 20 sec
Preprocessing - normalizing data ... data normalized in 9 min 53 sec
Preprocessing - running ANOVA ... excluded 22319 / 48367 genes in 1 min 11 sec
CPU times: user 1min 54s, sys: 9min 41s, total: 11min 36s
Wall time: 17min 46s


In [13]:
%%time
eso.compute(verbose=True)

Computing DET ... 
    esw ...
    empirical p-values ...
    esw_s ...
    finished in 7 min 24 sec
Computing EP ...
    esw ...
    empirical p-values ...
    esw_s ...
    finished in 0 min 0 sec
Computing GES ...
    esw ...
    empirical p-values ...
    esw_s ...
    finished in 0 min 10 sec
Computing NSI ...
    esw ...
    empirical p-values ...
    esw_s ...
    finished in 0 min 32 sec
Computing ESmu ...
    finished in 0 min 0 sec
Computing ESsd ...
    finished in 0 min 0 sec
Computed ['det.esw', 'det.esw_null', 'det.pvals', 'det.esw_s', 'ep.esw', 'ep.esw_null', 'ep.pvals', 'ep.esw_s', 'ges.esw', 'ges.esw_null', 'ges.pvals', 'ges.esw_s', 'nsi.esw', 'nsi.esw_null', 'nsi.pvals', 'nsi.esw_s', 'esmu', 'essd'].
CPU times: user 50.4 s, sys: 4min 21s, total: 5min 11s
Wall time: 8min 8s


In [14]:
eso.results['esmu']

Astro_L1-2_FGFR3_GFAP  Astro_L1-6_FGFR3_SLC14A1  \
gene                                                           
3.8-1.4                   0.000000                  0.000000   
A1BG                      0.000000                  0.000000   
A1CF                      0.000000                  0.000000   
A2M                       0.000000                  0.020191   
A2ML1                     0.172838                  0.325999   
...                            ...                       ...   
ZYG11B                    0.000000                  0.000000   
ZYX                       0.000000                  0.000000   
ZZEF1                     0.000000                  0.000000   
ZZZ3                      0.000000                  0.000000   
bA395L14.12               0.000000                  0.000000   

             Endo_L2-6_NOSTRIN  Exc_L2-3_LINC00507_FREM3  \
gene                                                       
3.8-1.4               0.000000                  0.000000   
A1BG                  0.000000                  0.116972   
A1CF                  0.000000                  0.006619   
A2M                   0.327568                  0.000000   
A2ML1                 0.000000                  0.110146   
...                        ...                       ...   
ZYG11B                0.000000                  0.000000   
ZYX                   0.000000                  0.034845   
ZZEF1                 0.000000                  0.050515   
ZZZ3                  0.000000                  0.055500   
bA395L14.12           0.000000                  0.000000   

             Exc_L2-4_LINC00507_GLP2R  Exc_L2_LAMP5_LTK  \
gene                                                      
3.8-1.4                      0.000000          0.000000   
A1BG                         0.000000          0.173146   
A1CF                         0.000000          0.000000   
A2M                          0.000000          0.000000   
A2ML1                        0.000000          0.000000   
...                               ...               ...   
ZYG11B                       0.032583          0.000000   
ZYX                          0.194453          0.163655   
ZZEF1                        0.000000          0.000000   
ZZZ3                         0.000000          0.000000   
bA395L14.12                  0.000000          0.000000   

             Exc_L3-4_RORB_CARM1P1  Exc_L3-5_RORB_COL22A1  Exc_L3-5_RORB_ESR1  \
gene                                                                            
3.8-1.4                   0.000000               0.000000            0.000000   
A1BG                      0.000000               0.137689            0.000000   
A1CF                      0.120175               0.000000            0.000000   
A2M                       0.000000               0.000000            0.000000   
A2ML1                     0.000000               0.000000            0.000000   
...                            ...                    ...                 ...   
ZYG11B                    0.000000               0.000000            0.021065   
ZYX                       0.000000               0.000000            0.000000   
ZZEF1                     0.000000               0.000000            0.000000   
ZZZ3                      0.000000               0.000000            0.000000   
bA395L14.12               0.000000               0.000000            0.000000   

             Exc_L3-5_RORB_FILIP1L  ...  Inh_L4-6_SST_GXYLT2  \
gene                                ...                        
3.8-1.4                        0.0  ...             0.000000   
A1BG                           0.0  ...             0.000000   
A1CF                           0.0  ...             0.000000   
A2M                            0.0  ...             0.000000   
A2ML1                          0.0  ...             0.154021   
...                            ...  ...                  ...   
ZYG11B                         0.0  ...             0.000000   
Z

In [16]:
cellex.utils.mapping.human_symbol_to_human_ens(eso.results['esmu'], drop_unmapped=True, verbose=True)

Mapping: human gene symbols --> human ensembl gene id's ...
30.52 pct of genes are unmapped ...
Removed 7949 unmapped genes ...


Using the gene column for gene names and then converting them to ensembl gene id's using 
```cellex.utils.mapping.human_symbol_to_human_ens``` results in 30.52 pct (7949) of genes being unmapped.

Using other methods including using the entrez_id column maps even less genes, so this method was still used.

In [17]:
eso.results['esmu']

Astro_L1-2_FGFR3_GFAP  Astro_L1-6_FGFR3_SLC14A1  \
gene                                                               
ENSG00000121410               0.000000                  0.000000   
ENSG00000148584               0.000000                  0.000000   
ENSG00000175899               0.000000                  0.020191   
ENSG00000166535               0.172838                  0.325999   
ENSG00000256069               0.000000                  0.000000   
...                                ...                       ...   
ENSG00000232242               0.000000                  0.000000   
ENSG00000162378               0.000000                  0.000000   
ENSG00000159840               0.000000                  0.000000   
ENSG00000074755               0.000000                  0.000000   
ENSG00000036549               0.000000                  0.000000   

                 Endo_L2-6_NOSTRIN  Exc_L2-3_LINC00507_FREM3  \
gene                                                           
ENSG00000121410           0.000000                  0.116972   
ENSG00000148584           0.000000                  0.006619   
ENSG00000175899           0.327568                  0.000000   
ENSG00000166535           0.000000                  0.110146   
ENSG00000256069           0.624466                  0.000000   
...                            ...                       ...   
ENSG00000232242           0.000000                  0.000000   
ENSG00000162378           0.000000                  0.000000   
ENSG00000159840           0.000000                  0.034845   
ENSG00000074755           0.000000                  0.050515   
ENSG00000036549           0.000000                  0.055500   

                 Exc_L2-4_LINC00507_GLP2R  Exc_L2_LAMP5_LTK  \
gene                                                          
ENSG00000121410                  0.000000          0.173146   
ENSG00000148584                  0.000000          0.000000   
ENSG00000175899                  0.000000          0.000000   
ENSG00000166535                  0.000000          0.000000   
ENSG00000256069                  0.000000          0.000000   
...                                   ...               ...   
ENSG00000232242                  0.000000          0.000000   
ENSG00000162378                  0.032583          0.000000   
ENSG00000159840                  0.194453          0.163655   
ENSG00000074755                  0.000000          0.000000   
ENSG00000036549                  0.000000          0.000000   

                 Exc_L3-4_RORB_CARM1P1  Exc_L3-5_RORB_COL22A1  \
gene                                                            
ENSG00000121410               0.000000               0.137689   
ENSG00000148584               0.120175               0.000000   
ENSG00000175899               0.000000               0.000000   
ENSG00000166535               0.000000               0.000000   
ENSG00000256069               0.000000               0.000000   
...                                ...                    ...   
ENSG00000232242               0.000000               0.000000   
ENSG00000162378               0.000000               0.000000   
ENSG00000159840               0.000000               0.000000   
ENSG00000074755               0.000000               0.000000   
ENSG00000036549               0.000000               0.000000   

                 Exc_L3-5_RORB_ESR1  Exc_L3-5_RORB_FILIP1L  ...  \
gene                                                        ...   
ENSG00000121410            0.000000                    0.0  ...   
ENSG00000148584            0.000000                    0.0  ...   
ENSG00000175899            0.000000                    0.0  ...   
ENSG00000166535            0.000000                    0.0  ...   
ENSG00000256069            0.000000                    0.0  ...   
...                             ...                    ...  ...   
ENSG00000232242            0.000000                    0.0  ...   
ENSG00000162378            0.0

In [18]:
eso.results["esmu"].to_csv("Allen_human_MTG.esmu.csv.gz")